GsaGH/Parameters/1_Properties/GsaMaterial.cs

```diff
@@ -479,6 +504,116 @@ public enum MatType {
       }
     }
 
+    private void ValidateApiMaterial() {
+      List<string> gradeNames = GetGradeNames(
+        MaterialType, SteelDesignCodeName, ConcreteDesignCodeName);
+      if (gradeNames.Contains(Name)) {
+        AnalysisMaterial a = CreateCodeAnalysisMaterial(Name);
+        if (AnalysisMaterialsAreEqual(a, AnalysisMaterial)) {
+          return;
+        }
+
+        _gradeName = Name;
+      } else {
+        var dict = new Dictionary<string, double>();
+        foreach (string gradeName in gradeNames) {
+          AnalysisMaterial a = CreateCodeAnalysisMaterial(gradeName);
+          if (AnalysisMaterialsAreEqual(a, AnalysisMaterial)) {
+            _gradeName = gradeName;
+            return;
+          }
+
+          dict[gradeName] = AnalysisMaterialsSimilarity(a, AnalysisMaterial);
+        }
+
+        _gradeName = dict.Aggregate((x, y) => x.Value > y.Value ? x : y).Key;
+      }
+    }
+
+    private double AnalysisMaterialsSimilarity(AnalysisMaterial a, AnalysisMaterial b) {
+      if (a == null || b == null) {
+        return 0;
+      }
+
+      double similarity = 0;
+      if (Math.Round(a.CoefficientOfThermalExpansion, 11) ==
+        Math.Round(b.CoefficientOfThermalExpansion, 11)) {
+        similarity += 1;
+      } else {
+        similarity += Math.Abs((a.CoefficientOfThermalExpansion - b.CoefficientOfThermalExpansion)
+          / b.CoefficientOfThermalExpansion);
+      }
+
+      if (Math.Round(a.Density, 11) == Math.Round(b.Density, 11)) {
+        similarity += 1;
+      } else {
+        similarity += Math.Abs((a.Density - b.Density) / b.Density);
+      }
+
+      if (Math.Round(a.ElasticModulus, 11) == Math.Round(b.ElasticModulus, 11)) {
+        similarity += 1;
+      } else {
+        similarity += Math.Abs((a.ElasticModulus - b.ElasticModulus) / b.ElasticModulus);
+      }
+
+      if (Math.Round(a.PoissonsRatio, 11) == Math.Round(b.PoissonsRatio, 11)) {
+        similarity += 1;
+      } else {
+        similarity += Math.Abs((a.PoissonsRatio - b.PoissonsRatio)/ b.PoissonsRatio);
+      }
+
+      return similarity;
+    }
+
+    private bool AnalysisMaterialsAreEqual(AnalysisMaterial a, AnalysisMaterial b) {
+      if (a == null || b == null) {
+        return false;
+      }
+
+      if (Math.Round(a.CoefficientOfThermalExpansion, 11) != 
```

Comment:
can be in one condition

Reply:
done: https://github.com/arup-group/GSA-Grasshopper/pull/434/commits/05d6ca1a9110d1a23895ed102c772d253ae854f4